main.go

```diff
@@ -114,60 +112,99 @@ func main() {
+	if serviceList.Consumer {
+		eventConsumer.Start(eventLoopContext, eventLoopDone, serviceIdentityValidated)
+	}
+
+	// Log non-fatal errors, without exiting
+	go func() {
+		var consumerErrors chan (error)
+		if serviceList.Consumer {
+			consumerErrors = syncConsumerGroup.Errors()
+		} else {
+			consumerErrors = make(chan error, 1)
+		}
+
+		var producerErrors chan (error)
+		if serviceList.DimensionExtractedProducer {
+			producerErrors = service.DimensionExtractedProducer.Errors()
+		} else {
+			producerErrors = make(chan error, 1)
+		}
+
+		select {
+		case consumerError := <-consumerErrors:
+			log.ErrorC("kafka consumer", consumerError, nil)
+		case producerError := <-producerErrors:
+			log.ErrorC("kafka producer", producerError, nil)
+		case apiError := <-apiErrors:
+			log.ErrorC("server error", apiError, nil)
+		}
+	}()
 
-	// block until a fatal error, signal or eventLoopDone - then proceed to shutdown
+	// Block until a fatal error occurs
 	select {
 	case <-eventLoopDone:
 		log.Debug("quitting after done was closed", nil)
 	case signal := <-signals:
 		log.Debug("quitting after os signal received", log.Data{"signal": signal})
-	case consumerError := <-syncConsumerGroup.Errors():
-		log.Error(fmt.Errorf("aborting consumer"), log.Data{"message_received": consumerError})
-	case producerError := <-service.DimensionExtractedProducer.Errors():
-		log.Error(fmt.Errorf("aborting producer"), log.Data{"message_received": producerError})
-	case <-apiErrors:
-		log.Error(fmt.Errorf("server error forcing shutdown"), nil)
 	}
 
 	// give the app `Timeout` seconds to close gracefully before killing it.
 	ctx, cancel := context.WithTimeout(context.Background(), cfg.GracefulShutdownTimeout)
 
 	go func() {
-		log.Debug("stopping kafka consumer listener", nil)
-		syncConsumerGroup.StopListeningToConsumer(ctx)
-		log.Debug("stopped kafka consumer listener", nil)
+
+		// If kafka consumer exists, stop listening to it. (Will close later)
+		if serviceList.Consumer {
+			log.Debug("stopping kafka consumer listener", nil)
+			syncConsumerGroup.StopListeningToConsumer(ctx)
+			log.Debug("stopped kafka consumer listener", nil)
+		}
+
 		eventLoopCancel()
 		<-eventLoopDone
+
+		// Close API
 		log.Debug("closing http server", nil)
 		if err := api.Close(ctx); err != nil {
 			log.ErrorC("failed to gracefully close http server", err, nil)
 		} else {
 			log.Debug("gracefully closed http server", nil)
 		}
-		log.Debug("closing dimension extracted kafka producer", nil)
-		service.DimensionExtractedProducer.Close(ctx)
-		log.Debug("closed dimension extracted kafka producer", nil)
 
-		log.Debug("closing down dimension extracted error producer", nil)
-		dimensionExtractedErrProducer.Close(ctx)
-		log.Debug("closed dimension extracted error producer", nil)
+		// If DimensionExtracted kafka producer exists, close it
+		if serviceList.DimensionExtractedProducer {
+			log.Debug("closing dimension extracted kafka producer", nil)
```

Comment:
Minor point - I know you wouldn't be aware of this but we have a logging standard where we don't have dynamic or variable dependant messages. 
You have the multiple messages for the same event but the only thing that is different is the producer being closed. If you make the message more general you can move the specific context to the logging params i.e.

```
log.Debug("closing kafka producer", log.Data{"producer": dimension extracted})
```